# CHB-MIT Dataset Explore

In [1]:
%matplotlib widget

## Check patients list

In [ ]:
import os

dataset_path = "./CHB-MIT"
patient_dirs = [d for d in os.listdir(dataset_path) if d.startswith("chb")]
patient_dirs.sort()
patient_dirs

['chb01',
 'chb02',
 'chb03',
 'chb04',
 'chb05',
 'chb06',
 'chb07',
 'chb08',
 'chb09',
 'chb10',
 'chb11',
 'chb12',
 'chb13',
 'chb14',
 'chb15',
 'chb16',
 'chb17',
 'chb18',
 'chb19',
 'chb20',
 'chb21',
 'chb22',
 'chb23',
 'chb24']

## Parse summary files

In [3]:
from typing import List
from utils import info

summary_info: List[info.PatientSummary] = info.parse_all_summary_files(dataset_path)
filtered_summary_info = info.filter_common_channels(summary_info)

Extracting EDF parameters from d:\KMUTT_Learnfile\2024-2\Project\CHB-MIT\chb24\chb24_01.edf...
EDF file detected
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 921599  =      0.000 ...  3599.996 secs...


d:\KMUTT_Learnfile\2024-2\Project\utils\info.py:217: RuntimeWarning: Channel names are not unique, found duplicates for: {'T8-P8'}. Applying running numbers for duplicates.
  raw = mne.io.read_raw_edf(os.path.join(dataset_path, "chb24", file["name"]), preload=True)


Extracting EDF parameters from d:\KMUTT_Learnfile\2024-2\Project\CHB-MIT\chb24\chb24_03.edf...
EDF file detected
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 921599  =      0.000 ...  3599.996 secs...


d:\KMUTT_Learnfile\2024-2\Project\utils\info.py:217: RuntimeWarning: Channel names are not unique, found duplicates for: {'T8-P8'}. Applying running numbers for duplicates.
  raw = mne.io.read_raw_edf(os.path.join(dataset_path, "chb24", file["name"]), preload=True)


Extracting EDF parameters from d:\KMUTT_Learnfile\2024-2\Project\CHB-MIT\chb24\chb24_04.edf...
EDF file detected
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 921599  =      0.000 ...  3599.996 secs...


d:\KMUTT_Learnfile\2024-2\Project\utils\info.py:217: RuntimeWarning: Channel names are not unique, found duplicates for: {'T8-P8'}. Applying running numbers for duplicates.
  raw = mne.io.read_raw_edf(os.path.join(dataset_path, "chb24", file["name"]), preload=True)


Extracting EDF parameters from d:\KMUTT_Learnfile\2024-2\Project\CHB-MIT\chb24\chb24_06.edf...
EDF file detected
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 921599  =      0.000 ...  3599.996 secs...


d:\KMUTT_Learnfile\2024-2\Project\utils\info.py:217: RuntimeWarning: Channel names are not unique, found duplicates for: {'T8-P8'}. Applying running numbers for duplicates.
  raw = mne.io.read_raw_edf(os.path.join(dataset_path, "chb24", file["name"]), preload=True)


Extracting EDF parameters from d:\KMUTT_Learnfile\2024-2\Project\CHB-MIT\chb24\chb24_07.edf...
EDF file detected
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 921599  =      0.000 ...  3599.996 secs...


d:\KMUTT_Learnfile\2024-2\Project\utils\info.py:217: RuntimeWarning: Channel names are not unique, found duplicates for: {'T8-P8'}. Applying running numbers for duplicates.
  raw = mne.io.read_raw_edf(os.path.join(dataset_path, "chb24", file["name"]), preload=True)


Extracting EDF parameters from d:\KMUTT_Learnfile\2024-2\Project\CHB-MIT\chb24\chb24_09.edf...
EDF file detected
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 921599  =      0.000 ...  3599.996 secs...


d:\KMUTT_Learnfile\2024-2\Project\utils\info.py:217: RuntimeWarning: Channel names are not unique, found duplicates for: {'T8-P8'}. Applying running numbers for duplicates.
  raw = mne.io.read_raw_edf(os.path.join(dataset_path, "chb24", file["name"]), preload=True)


Extracting EDF parameters from d:\KMUTT_Learnfile\2024-2\Project\CHB-MIT\chb24\chb24_11.edf...
EDF file detected
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 921599  =      0.000 ...  3599.996 secs...


d:\KMUTT_Learnfile\2024-2\Project\utils\info.py:217: RuntimeWarning: Channel names are not unique, found duplicates for: {'T8-P8'}. Applying running numbers for duplicates.
  raw = mne.io.read_raw_edf(os.path.join(dataset_path, "chb24", file["name"]), preload=True)


Extracting EDF parameters from d:\KMUTT_Learnfile\2024-2\Project\CHB-MIT\chb24\chb24_13.edf...
EDF file detected
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 921599  =      0.000 ...  3599.996 secs...


d:\KMUTT_Learnfile\2024-2\Project\utils\info.py:217: RuntimeWarning: Channel names are not unique, found duplicates for: {'T8-P8'}. Applying running numbers for duplicates.
  raw = mne.io.read_raw_edf(os.path.join(dataset_path, "chb24", file["name"]), preload=True)


Extracting EDF parameters from d:\KMUTT_Learnfile\2024-2\Project\CHB-MIT\chb24\chb24_14.edf...
EDF file detected
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 921599  =      0.000 ...  3599.996 secs...


d:\KMUTT_Learnfile\2024-2\Project\utils\info.py:217: RuntimeWarning: Channel names are not unique, found duplicates for: {'T8-P8'}. Applying running numbers for duplicates.
  raw = mne.io.read_raw_edf(os.path.join(dataset_path, "chb24", file["name"]), preload=True)


Extracting EDF parameters from d:\KMUTT_Learnfile\2024-2\Project\CHB-MIT\chb24\chb24_15.edf...
EDF file detected
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 921599  =      0.000 ...  3599.996 secs...


d:\KMUTT_Learnfile\2024-2\Project\utils\info.py:217: RuntimeWarning: Channel names are not unique, found duplicates for: {'T8-P8'}. Applying running numbers for duplicates.
  raw = mne.io.read_raw_edf(os.path.join(dataset_path, "chb24", file["name"]), preload=True)


Extracting EDF parameters from d:\KMUTT_Learnfile\2024-2\Project\CHB-MIT\chb24\chb24_17.edf...
EDF file detected
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 921599  =      0.000 ...  3599.996 secs...


d:\KMUTT_Learnfile\2024-2\Project\utils\info.py:217: RuntimeWarning: Channel names are not unique, found duplicates for: {'T8-P8'}. Applying running numbers for duplicates.
  raw = mne.io.read_raw_edf(os.path.join(dataset_path, "chb24", file["name"]), preload=True)


Extracting EDF parameters from d:\KMUTT_Learnfile\2024-2\Project\CHB-MIT\chb24\chb24_21.edf...
EDF file detected
Setting channel info structure...
Creating raw.info structure...
Reading 0 ... 921599  =      0.000 ...  3599.996 secs...


d:\KMUTT_Learnfile\2024-2\Project\utils\info.py:217: RuntimeWarning: Channel names are not unique, found duplicates for: {'T8-P8'}. Applying running numbers for duplicates.
  raw = mne.io.read_raw_edf(os.path.join(dataset_path, "chb24", file["name"]), preload=True)


Let check the result

In [5]:
from pprint import pprint

pprint(filtered_summary_info[0], width=80, indent=2)

{ 'channels_list': [ [ 'FP1-F7',
                       'F7-T7',
                       'T7-P7',
                       'P7-O1',
                       'FP1-F3',
                       'F3-C3',
                       'C3-P3',
                       'P3-O1',
                       'FP2-F4',
                       'F4-C4',
                       'C4-P4',
                       'P4-O2',
                       'FP2-F8',
                       'F8-T8',
                       'T8-P8',
                       'P8-O2',
                       'FZ-CZ',
                       'CZ-PZ',
                       'P7-T7',
                       'T7-FT9',
                       'FT9-FT10',
                       'FT10-T8',
                       'T8-P8']],
  'files': [ { 'channels_set_idx': 0,
               'end_time': datetime.datetime(2075, 3, 10, 12, 42, 54),
               'name': 'chb01_01.edf',
               'seizures': [],
               'start_time': datetime.datetime(2075, 3, 10, 11, 42, 54)},

In [ ]:
import pandas as pd

pd.DataFrame(filtered_summary_info[23]['files'])

name  channels_set_idx                start_time  \
0   chb24_01.edf                 0 2074-01-25 16:26:28+00:00   
1   chb24_03.edf                 0 2074-01-25 18:26:39+00:00   
2   chb24_04.edf                 0 2074-01-25 19:26:46+00:00   
3   chb24_06.edf                 0 2074-01-25 21:27:00+00:00   
4   chb24_07.edf                 0 2074-01-25 22:27:08+00:00   
5   chb24_09.edf                 0 2074-01-26 00:27:22+00:00   
6   chb24_11.edf                 0 2074-01-26 02:27:36+00:00   
7   chb24_13.edf                 0 2074-01-26 04:27:50+00:00   
8   chb24_14.edf                 0 2074-01-26 05:27:57+00:00   
9   chb24_15.edf                 0 2074-01-26 06:28:05+00:00   
10  chb24_17.edf                 0 2074-01-26 08:28:19+00:00   
11  chb24_21.edf                 0 2074-01-26 12:28:50+00:00   

                    end_time  \
0  2074-01-25 17:26:28+00:00   
1  2074-01-25 19:26:39+00:00   
2  2074-01-25 20:26:46+00:00   
3  2074-01-25 22:27:00+00:00   
4  2074-01-25 23:27:08+00:00   
5  2074-01-26 01:27:22+00:00   
6  2074-01-26 03:27:36+00:00   
7  2074-01-26 05:27:50+00:00   
8  2074-01-26 06:27:57+00:00   
9  2074-01-26 07:28:05+00:00   
10 2074-01-26 09:28:19+00:00   
11 2074-01-26 13:28:50+00:00   

                                             seizures  
0   [{'start_time': 480, 'end_time': 505}, {'start...  
1   [{'start_time': 231, 'end_time': 260}, {'start...  
2   [{'start_time': 1088, 'end_time': 1120}, {'sta...  
3            [{'start_time': 1229, 'end_time': 1253}]  
4                [{'start_time': 38, 'end_time': 60}]  
5            [{'start_time': 1745, 'end_time': 1764}]  
6            [{'start_time': 3527, 'end_time': 3597}]  
7            [{'start_time': 3288, 'end_time': 3304}]  
8            [{'start_time': 1939, 'end_time': 1966}]  
9            [{'start_time': 3552, 'end_time': 3569}]  
10           [{'start_time': 3515, 'end_time': 3581}]  
11           [{'start_time': 2804, 'end_time': 2872}]

# Total Seizures in dataset

In [7]:
import pandas as pd

seizure_count = {}

for i, patient_info in enumerate(filtered_summary_info):
    patient_id = f"chb{(i + 1):02}"
    for file in patient_info['files']:
        for seizure in file["seizures"]:
            seizure_count[patient_id] = seizure_count.get(patient_id, 0) + 1

sezure_count = pd.DataFrame.from_dict(seizure_count, orient='index', columns=['seizure_count'])

display(sezure_count)
display(sezure_count.sum())

seizure_count
chb01              7
chb02              3
chb03              7
chb04              4
chb05              5
chb06             10
chb07              3
chb08              5
chb09              4
chb10              7
chb11              3
chb12             27
chb13             10
chb14              8
chb15             20
chb16              8
chb17              3
chb18              6
chb19              3
chb20              8
chb21              4
chb22              3
chb23              7
chb24             16

seizure_count    181
dtype: int64

In [8]:
COMMON_CHANNELS = [
    "C3-P3",
    "C4-P4",
    "CZ-PZ",
    "F3-C3",
    "F4-C4",
    "F7-T7",
    "F8-T8",
    "FP1-F3",
    "FP1-F7",
    "FP2-F4",
    "FP2-F8",
    "FT10-T8",
    "FT9-FT10",
    "FZ-CZ",
    "P3-O1",
    "P4-O2",
    "P7-O1",
    "P7-T7",
    "P8-O2",
    "T7-FT9",
    "T7-P7",
    "T8-P8",
]

## Exclude abnormal data
We have checking each records and found that most of the record using this `"C3-P3","C4-P4","CZ-PZ","F3-C3","F4-C4","F7-T7","F8-T8","FP1-F3","FP1-F7","FP2-F4","FP2-F8","FT10-T8","FT9-FT10","FZ-CZ","P3-O1","P4-O2","P7-O1","P7-T7","P8-O2","T7-FT9","T7-P7","T8-P8"` excepts the `chb12_27.edf`, `chb12_28.edf` and `chb12_29.edf` that didn't use this channels configuration. But we already filter out the files those are not use the common channels already.

But for making sure, we will check which files those contain seizure don't contain all channels in common channels

In [9]:
from utils.constant import COMMON_CHANNELS

channels_set = set(COMMON_CHANNELS)

for i, patient_info in enumerate(filtered_summary_info):
    for file in patient_info['files']:
        if file["name"] in ["chb12_27.edf", "chb12_28.edf", "chb12_29.edf"]:
            continue
        
        current_file_channels = set(patient_info["channels_list"][file["channels_set_idx"]])
        missing_channels = channels_set - current_file_channels
        
        if (len(file["seizures"]) > 0) and len(missing_channels) > 0:
            print(f"Missing channels in {file['name']}: {missing_channels}, Seizure Count: {len(file['seizures'])}")

## Checking the total usable ictal class

In [10]:
from utils import dataset


ictal_list = dataset.list_ictal(filtered_summary_info)
ictal_df = pd.DataFrame(ictal_list)
ictal_df

file  start_time  end_time  total_time  total_windows
0    chb01_03.edf        2996      3036          40              9
1    chb01_04.edf        1467      1494          27              5
2    chb01_15.edf        1732      1772          40              9
3    chb01_16.edf        1015      1066          51             11
4    chb01_18.edf        1720      1810          90             21
..            ...         ...       ...         ...            ...
176  chb24_13.edf        3288      3304          16              3
177  chb24_14.edf        1939      1966          27              5
178  chb24_15.edf        3552      3569          17              3
179  chb24_17.edf        3515      3581          66             15
180  chb24_21.edf        2804      2872          68             16

[181 rows x 5 columns]

Now we can check how many windows we got for ictal class

In [11]:
ictal_df.sum()

file             chb01_03.edfchb01_04.edfchb01_15.edfchb01_16.e...
start_time                                                  473011
end_time                                                    484026
total_time                                                   11015
total_windows                                                 2509
dtype: object

## Checking the total usable interictal class

For the interictal, it is more tricky to choose as we need to consider the ictal class. If we choosing it too close to the ictal class it might capture the pre-ictal phrase. To be able to safely get the interictal class we need to get the item that in range before or after the ictal several hours.

In [12]:
interictal_list = dataset.list_interictal(filtered_summary_info, ictal_list)
interictal_df = pd.DataFrame(interictal_list)
interictal_df

file  total_time  total_windows
0    chb01_09.edf        3600            899
1    chb01_10.edf        3600            899
2    chb01_36.edf        3600            899
3    chb01_37.edf        3600            899
4    chb01_38.edf        3600            899
..            ...         ...            ...
199  chb22_15.edf        3600            899
200  chb22_30.edf        3600            899
201  chb22_54.edf        3600            899
202  chb23_19.edf       14400           3599
203  chb23_20.edf        5009           1251

[204 rows x 3 columns]

In [13]:
interictal_df.sum()

file             chb01_09.edfchb01_10.edfchb01_36.edfchb01_37.e...
total_time                                                 1140626
total_windows                                               284944
dtype: object

## Prepare dataset

From info above we can contruct the dataset that contain ictal and interictal class that both class has 2509 samples. And since the interictal class is for much more that the ictal class we can randomly select it to match the ictal class. Since we need 2509 windows we can roughly select 9 windows from each files and randomly discard the remainder.

In [ ]:
dataset.save_ictal("./CHB-MIT", ictal_list)

In [ ]:
dataset.save_interictal("./CHB-MIT", interictal_list)